src/daft-functions/src/coalesce.rs

```diff
@@ -0,0 +1,191 @@
+
+use common_error::{DaftError, DaftResult};
+use daft_core::{
+    prelude::{BooleanArray, DaftLogical, DataType, Field, Schema},
+    series::{IntoSeries, Series},
+};
+use daft_dsl::{functions::ScalarUDF, ExprRef};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct Coalesce {}
+
+#[typetag::serde]
+impl ScalarUDF for Coalesce {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &'static str {
+        "coalesce"
+    }
+
+    fn to_field(&self, inputs: &[ExprRef], schema: &Schema) -> DaftResult<Field> {
+        fn ensure_valid_dtype(Field { dtype, .. }: &Field) -> DaftResult<()> {
+            match dtype {
+                DataType::Boolean | DataType::Utf8 => Ok(()),
+                dt if dt.is_physical() || dt.is_primitive() => Ok(()),
+
+                dt if dt.is_list() | dt.is_nested() => {
+                    Err(DaftError::not_implemented("coalesce for nested datatypes"))
+                }
+                other => Err(DaftError::ComputeError(format!(
+                    "Unsupported data type for coalesce: {:?}",
+                    other
+                ))),
+            }
+        }
+
+        match inputs {
+            [] => Err(DaftError::SchemaMismatch(
+                "Expected at least 1 input args, got 0".to_string(),
+            )),
+            [input] => {
+                let input_field = input.to_field(schema)?;
+                ensure_valid_dtype(&input_field)?;
+                Ok(input_field)
+            }
+
+            _ => {
+                let first_field = inputs[0].to_field(schema)?;
+                ensure_valid_dtype(&first_field)?;
+
+                for input in inputs {
+                    if input.to_field(schema)?.dtype != first_field.dtype {
+                        return Err(DaftError::SchemaMismatch(
+                            "All input fields must have the same data type".to_string(),
+                        ));
+                    }
+                }
+                Ok(first_field)
+            }
+        }
+    }
+
+    fn evaluate(&self, inputs: &[Series]) -> DaftResult<Series> {
+        match inputs.len() {
+            0 => Err(DaftError::ComputeError("No inputs provided".to_string())),
+            1 => Ok(inputs[0].clone()),
+            _ => {
+                let name = inputs[0].name();
+                let dtype = inputs[0].data_type();
+                let len = inputs[0].len();
+                // the first input is not null, so no work to do
+                if inputs[0].validity().is_none() {
+                    return Ok(inputs[0].clone());
+                }
+
+                let mut current_value = Series::full_null(name, dtype, len).to_arrow();
+                let remainder = BooleanArray::from_values(name, vec![true; len].into_iter());
+                let mut remainder = remainder.into_series();
+
+                for input in inputs {
```

Comment:
an alternative implementation you could do is create a `GrowableArray` from all the input series and then select the index that you want to append. this way you can do this 1 pass over the data.

Reply:
https://github.com/Eventual-Inc/Daft/blob/75ad85a9eeee13603cfc5a158d87a1d141e785cc/src/daft-core/src/array/growable/mod.rs#L40

Reply:
not sure I follow here. Could we follow up with this optimization in a separate PR?